Unable to create new project for Sponge/Forge plugin

Hey team, unable to create a new project here. Getting the following error:
`java.lang.IllegalArgumentException: Failed to parse version part as integer: 0-SNAPSHOT`

How to reproduce issue:
![image](https://user-images.githubusercontent.com/3460315/34424565-77bb5e50-ec89-11e7-8328-a462610905da.png)
![image](https://user-images.githubusercontent.com/3460315/34424570-7de8238a-ec89-11e7-9228-11f2b703bdfe.png)
![image](https://user-images.githubusercontent.com/3460315/34424581-8c20b11a-ec89-11e7-9d6e-ad42f50c28c1.png)
![image](https://user-images.githubusercontent.com/3460315/34424586-9b62e6de-ec89-11e7-9666-c670ff9f002c.png)
![image](https://user-images.githubusercontent.com/3460315/34424588-a62865bc-ec89-11e7-8f48-5b2817d2c56e.png)


```
java.lang.IllegalArgumentException: Failed to parse version part as integer: 0-SNAPSHOT
java.lang.RuntimeException: java.lang.IllegalArgumentException: Failed to parse version part as integer: 0-SNAPSHOT
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:181)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:661)
	at com.demonwav.mcdev.buildsystem.gradle.GradleBuildSystem.handleForgeCreate(GradleBuildSystem.kt:488)
	at com.demonwav.mcdev.buildsystem.gradle.GradleBuildSystem.create(GradleBuildSystem.kt:74)
	at com.demonwav.mcdev.creator.MinecraftProjectCreator$doSingleModuleCreate$1.run(MinecraftProjectCreator.kt:75)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:713)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:157)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:543)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:488)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:144)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl$2.run(ProgressManagerImpl.java:165)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:315)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Failed to parse version part as integer: 0-SNAPSHOT
	at com.demonwav.mcdev.util.SemanticVersion$Companion$parse$1.invoke(SemanticVersion.kt:54)
	at com.demonwav.mcdev.util.SemanticVersion$Companion.parse(SemanticVersion.kt:67)
	at com.demonwav.mcdev.platform.forge.ForgeTemplate.applyBuildGradleTemplate(ForgeTemplate.kt:38)
	at com.demonwav.mcdev.buildsystem.gradle.GradleBuildSystem$handleForgeCreate$$inlined$runWriteTask$2$1.run(Util.kt:147)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1003)
	at com.demonwav.mcdev.buildsystem.gradle.GradleBuildSystem$handleForgeCreate$$inlined$runWriteTask$2.run(Util.kt:145)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:315)
	at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:158)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.a(LaterInvocator.java:424)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:407)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:762)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:732)
	at com.intellij.ide.IdeEventQueue.h(IdeEventQueue.java:822)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:650)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:366)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

--- a/src/main/kotlin/com/demonwav/mcdev/creator/ForgeProjectSettingsWizard.kt
+++ b/src/main/kotlin/com/demonwav/mcdev/creator/ForgeProjectSettingsWizard.kt
@@ -202,7 +202,7 @@ class ForgeProjectSettingsWizard(private val creator: MinecraftProjectCreator) :
         if (fullVersion != null) {
             settings!!.forgeVersion = fullVersion
         }
-        settings!!.mcVersion = minecraftVersionBox.selectedItem as String
+        settings!!.mcVersion = version!!
     }
 
     fun error() {

--- a/src/main/kotlin/com/demonwav/mcdev/util/SemanticVersion.kt
+++ b/src/main/kotlin/com/demonwav/mcdev/util/SemanticVersion.kt
@@ -10,7 +10,6 @@
 
 package com.demonwav.mcdev.util
 
-import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.PreReleasePart
 import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.ReleasePart
 import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.TextPart
 
@@ -33,6 +32,8 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
 
     override fun hashCode() = parts.hashCode()
 
+    override fun toString() = versionString
+
     companion object {
         /**
          * The higher the priority value associated with a certain part, the higher it is ranked in comparisons.
@@ -40,12 +41,17 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
          */
         val TEXT_PRIORITIES = mapOf(
             "snapshot" to 0,
-            "rc" to 1
+            "rc" to 1,
+            "pre" to 1
         )
+        /**
+         * All separators allowed between a number and a modifier (i.e. (numbered) text part).
+         */
+        val SEPARATORS = listOf('-', '_')
 
         /**
          * Parses a version string into a comparable representation.
-         * @throws IllegalArgumentException if any part of the version string cannot be parsed as integer or split into pre-release parts.
+         * @throws IllegalArgumentException if any part of the version string cannot be parsed as integer or split into text parts.
          */
         fun parse(value: String): SemanticVersion {
             fun parseInt(part: String): Int =
@@ -54,24 +60,25 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
                 else
                     throw IllegalArgumentException("Failed to parse version part as integer: $part")
 
-            val parts = value.split('.').map {
-                if (it.contains("_pre")) {
-                    // There have been cases of Forge builds for MC pre-releases (1.7.10_pre4)
-                    // We're consuming the 10_pre4 and extracting 10 and 4 from it
-                    val subParts = it.split("_pre")
-                    if (subParts.size == 2) {
-                        PreReleasePart(parseInt(subParts[0]), parseInt(subParts[1]))
-                    } else {
-                        throw IllegalArgumentException("Failed to split pre-release version part into two numbers: $it")
-                    }
-                } else if (it.contains("-")) {
-                    // We might consume snapshot versions as well
-                    // Those are generally rated the lowest in comparisons
-                    // As a simplification, treat anything separated by a dash like that
-                    val subParts = it.split("-", limit = 2)
-                    TextPart(parseInt(subParts[0]), subParts[1])
+            // We need to support pre-releases/RCs and snapshots as well
+            fun parseTextPart(subParts: List<String>, separator: Char): VersionPart =
+                if (subParts.size == 2) {
+                    val version = parseInt(subParts[0])
+                    val (text, number) = subParts[1].span { !it.isDigit() }
+                    // Pure text parts always are considered older than numbered ones
+                    // Since we don't handle negative version numbers, -1 is guaranteed to be smaller than any numbered part
+                    val versionNumber = if (number.isEmpty()) -1 else parseInt(number)
+                    TextPart(version, separator, text, versionNumber)
                 } else {
-                    ReleasePart(parseInt(it))
+                    throw IllegalArgumentException("Failed to split text version part into two: ${subParts.first()}$separator")
+                }
+
+            val parts = value.split('.').map { part ->
+                val separator = SEPARATORS.find { it in part }
+                if (separator != null) {
+                    parseTextPart(part.split(separator, limit = 2), separator)
+                } else {
+                    ReleasePart(parseInt(part))
                 }
             }
             return SemanticVersion(parts)
@@ -86,32 +93,24 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
                 override fun compareTo(other: VersionPart) =
                     when (other) {
                         is ReleasePart -> version - other.version
-                        is PreReleasePart -> if (version != other.version) version - other.version else 1
                         is TextPart -> if (version != other.version) version - other.version else 1
                     }
             }
 
-            data class PreReleasePart(val version: Int, val pre: Int) : VersionPart() {
-                override val versionString = "${version}_pre$pre"
-
-                override fun compareTo(other: VersionPart) =
-                    when (other) {
-                        is ReleasePart -> if (version != other.version) version - other.version else -1
-                        is PreReleasePart -> if (version != other.version) version - other.version else pre - other.pre
-                        is TextPart -> if (version != other.version) version - other.version else 1
-                    }
-            }
-
-            data class TextPart(val version: Int, val text: String) : VersionPart() {
+            data class TextPart(val version: Int, val separator: Char, val text: String, val number: Int) : VersionPart() {
                 private val priority = TEXT_PRIORITIES[text.toLowerCase()] ?: -1
 
-                override val versionString = "$version-$text"
+                override val versionString = "$version$separator$text${if (number == -1) "" else number.toString()}"
 
                 override fun compareTo(other: VersionPart) =
                     when (other) {
                         is ReleasePart -> if (version != other.version) version - other.version else -1
-                        is PreReleasePart -> if (version != other.version) version - other.version else -1
-                        is TextPart -> if (version != other.version) version - other.version else priority - other.priority
+                        is TextPart ->
+                            when {
+                                version != other.version -> version - other.version
+                                text != other.text -> priority - other.priority
+                                else -> number - other.number
+                            }
                     }
             }
         }

--- a/src/main/kotlin/com/demonwav/mcdev/util/SemanticVersion.kt
+++ b/src/main/kotlin/com/demonwav/mcdev/util/SemanticVersion.kt
@@ -12,6 +12,7 @@ package com.demonwav.mcdev.util
 
 import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.PreReleasePart
 import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.ReleasePart
+import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.TextPart
 
 /**
  * Represents a comparable and generalised "semantic version".
@@ -24,7 +25,7 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
     override fun compareTo(other: SemanticVersion): Int {
         // Zipping limits the compared parts to the shorter version, then we perform a component-wise comparison
         // Short-circuits if any component of this version is smaller/older
-        val result = parts.zip(other.parts).fold(0) { acc, (a, b) -> if (acc == -1) acc else a.compareTo(b) }
+        val result = parts.zip(other.parts).fold(0) { acc, (a, b) -> if (acc != 0) acc else a.compareTo(b) }
         // When all the parts are equal, the longer version wins
         // Generally speaking, 1.0 is considered older than 1.0.1 (see MC 1.12 vs 1.12.1)
         if (parts.size != other.parts.size && result == 0) {
@@ -42,6 +43,15 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
     override fun hashCode() = parts.hashCode()
 
     companion object {
+        /**
+         * The higher the priority value associated with a certain part, the higher it is ranked in comparisons.
+         * Unknown parts will always be "rated" lower and equal among each other.
+         */
+        val TEXT_PRIORITIES = mapOf(
+            "snapshot" to 0,
+            "rc" to 1
+        )
+
         /**
          * Parses a version string into a comparable representation.
          * @throws IllegalArgumentException if any part of the version string cannot be parsed as integer or split into pre-release parts.
@@ -63,6 +73,12 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
                     } else {
                         throw IllegalArgumentException("Failed to split pre-release version part into two numbers: $it")
                     }
+                } else if (it.contains("-")) {
+                    // We might consume snapshot versions as well
+                    // Those are generally rated the lowest in comparisons
+                    // As a simplification, treat anything separated by a dash like that
+                    val subParts = it.split("-", limit = 2)
+                    TextPart(parseInt(subParts[0]), subParts[1])
                 } else {
                     ReleasePart(parseInt(it))
                 }
@@ -78,8 +94,9 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
 
                 override fun compareTo(other: VersionPart) =
                     when (other) {
-                        is PreReleasePart -> if (version != other.version) version - other.version else 1
                         is ReleasePart -> version - other.version
+                        is PreReleasePart -> if (version != other.version) version - other.version else 1
+                        is TextPart -> if (version != other.version) version - other.version else 1
                     }
             }
 
@@ -88,8 +105,22 @@ class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion
 
                 override fun compareTo(other: VersionPart) =
                     when (other) {
-                        is PreReleasePart -> if (version != other.version) version - other.version else pre - other.pre
                         is ReleasePart -> if (version != other.version) version - other.version else -1
+                        is PreReleasePart -> if (version != other.version) version - other.version else pre - other.pre
+                        is TextPart -> if (version != other.version) version - other.version else 1
+                    }
+            }
+
+            data class TextPart(val version: Int, val text: String) : VersionPart() {
+                private val priority = TEXT_PRIORITIES[text.toLowerCase()] ?: -1
+
+                override val versionString = "$version-$text"
+
+                override fun compareTo(other: VersionPart) =
+                    when (other) {
+                        is ReleasePart -> if (version != other.version) version - other.version else -1
+                        is PreReleasePart -> if (version != other.version) version - other.version else -1
+                        is TextPart -> if (version != other.version) version - other.version else priority - other.priority
                     }
             }
         }

--- a/src/main/kotlin/com/demonwav/mcdev/util/SemanticVersion.kt
+++ b/src/main/kotlin/com/demonwav/mcdev/util/SemanticVersion.kt
@@ -22,17 +22,8 @@ import com.demonwav.mcdev.util.SemanticVersion.Companion.VersionPart.TextPart
 class SemanticVersion(val parts: List<VersionPart>) : Comparable<SemanticVersion> {
     val versionString = parts.joinToString(".") { it.versionString }
 
-    override fun compareTo(other: SemanticVersion): Int {
-        // Zipping limits the compared parts to the shorter version, then we perform a component-wise comparison
-        // Short-circuits if any component of this version is smaller/older
-        val result = parts.zip(other.parts).fold(0) { acc, (a, b) -> if (acc != 0) acc else a.compareTo(b) }
-        // When all the parts are equal, the longer version wins
-        // Generally speaking, 1.0 is considered older than 1.0.1 (see MC 1.12 vs 1.12.1)
-        if (parts.size != other.parts.size && result == 0) {
-            return parts.size - other.parts.size
-        }
-        return result
-    }
+    override fun compareTo(other: SemanticVersion): Int =
+        naturalOrder<VersionPart>().lexicographical().compare(parts, other.parts)
 
     override fun equals(other: Any?) =
         when (other) {

--- a/src/main/kotlin/com/demonwav/mcdev/util/sorting.kt
+++ b/src/main/kotlin/com/demonwav/mcdev/util/sorting.kt
@@ -10,62 +10,27 @@
 
 package com.demonwav.mcdev.util
 
-import java.util.ArrayList
-import java.util.Arrays
 import java.util.Comparator
-import java.util.stream.Collectors
-import java.util.stream.Stream
 
-val LEXICOGRAPHICAL_ORDER: Comparator<IntArray> = Comparator { one, two ->
-    val length = Math.min(one.size, two.size)
-    for (i in 0 until length) {
-        val first = one[i]
-        val second = two[i]
-
-        if (first < second) {
-            return@Comparator -1
-        } else if (second < first) {
-            return@Comparator 1
+fun <T> Comparator<in T>.lexicographical(): Comparator<in Iterable<T>> =
+    Comparator { left, right ->
+        // Zipping limits the compared parts to the shorter list, then we perform a component-wise comparison
+        // Short-circuits if any component of the left side is smaller or greater
+        left.zip(right).fold(0) { acc, (a, b) -> if (acc != 0) return@Comparator acc else this.compare(a, b) }.let {
+            // When all the parts are equal, the longer list wins (is greater)
+            if (it == 0) {
+                left.count() - right.count()
+            } else {
+                // Still required if the last part was not equal, the short-circuiting does not cover that
+                it
+            }
         }
     }
-
-    // We've got here so they are now equal, if one has more then they are not equal
-    if (one.size < two.size) {
-        return@Comparator -1
-    } else if (two.size < one.size) {
-        return@Comparator 1
-    }
-    // They are the same
-    return@Comparator 0
-}
-
-val REVERSE_LEXICOGRAPHICAL_ORDER: Comparator<IntArray> = LEXICOGRAPHICAL_ORDER.reversed()
 
 /**
  * This is the lowest version value we will let users choose, to make our lives easier.
  */
-private val ARRAY_1_8_8 = intArrayOf(1, 8, 8)
+private val MC_1_8_8 = SemanticVersion.parse("1.8.8")
 
-fun sortVersions(versions: Collection<*>): List<String> {
-    // Populate a list of the keys (and cast them to String) so they can be sorted
-    val list = ArrayList<String>(versions.size)
-    list.addAll(versions.stream().map(Any?::toString).collect(Collectors.toList<String>()))
-
-    // We map each version string (1.2, 1.9.4, 1.10, etc) to an array of integers {1, 2}, {1, 9, 4}, {1, 10} so we
-    // can lexicographically order them. We throw out the odd-balls in the process (like 1.10-pre4)
-    val intList = list.stream().distinct().mapNotNull { s ->
-        try {
-            return@mapNotNull Stream.of(*s.split("\\.".toRegex()).dropLastWhile(String::isEmpty).toTypedArray())
-                .mapToInt { Integer.parseInt(it) }.toArray()
-        } catch (e: NumberFormatException) {
-            return@mapNotNull null
-        }
-    }.collect(Collectors.toCollection { ArrayList<IntArray>() })
-
-    // Sort them correctly
-    intList.sortWith(REVERSE_LEXICOGRAPHICAL_ORDER)
-    intList.removeIf { ints -> LEXICOGRAPHICAL_ORDER.compare(ints, ARRAY_1_8_8) < 0 }
-
-    return intList.stream().map{ ints -> Arrays.stream(ints).mapToObj(Int::toString).collect(Collectors.joining(".")) }
-        .collect(Collectors.toList<String>())
-}
+fun sortVersions(versions: Collection<*>): List<String> =
+    versions.map(Any?::toString).map(SemanticVersion.Companion::parse).sortedDescending().filter { it >= MC_1_8_8 }.map { it.toString() }

--- a/src/main/kotlin/com/demonwav/mcdev/util/utils.kt
+++ b/src/main/kotlin/com/demonwav/mcdev/util/utils.kt
@@ -146,3 +146,13 @@ val gson = Gson()
 inline fun <reified T : Any> Gson.fromJson(text: String): T = fromJson(text, object : TypeToken<T>() {}.type)
 
 fun <K> Map<K, *>.containsAllKeys(vararg keys: K) = keys.all { this.containsKey(it) }
+
+/**
+ * Splits a string into the longest prefix matching a predicate and the corresponding suffix *not* matching.
+ *
+ * Note: Name inspired by Scala.
+ */
+inline fun String.span(predicate: (Char) -> Boolean): Pair<String, String> {
+    val prefix = takeWhile(predicate)
+    return prefix to drop(prefix.length)
+}

